### PR TITLE
Add error-prone compiler and fix its found issues.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: okhttp-tests/src/test/java/okhttp3/internal/platform/OptionalMethodTest.java [hunk 1]
```diff
@@ -309,9 +311,10 @@ public void invokeOptionalNonPublic() throws Exception {
       OptionalMethod<T> optionalMethod, T base, Object... args) throws Exception {
     try {
       optionalMethod.invoke(base, args);
-      fail();
     } catch (Error expected) {
+      return;
     }
+    fail();
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/okhttp3/internal/platform/OptionalMethodTest.java]: yep

**Comment by JakeWharton** [comment]
Looks like error-prone only runs on JDK 8. Are we okay dropping openjdk7 as a compilation target?

**Comment by JakeWharton** [comment]
As a bonus, though, error-prone is based on JDK 9 which means it should support the new `-release` flag for targeting older platforms...

**Comment by swankjesse** [comment]
I'm happy to bump our minimum desktop JVM to JDK 8. Does that impact IntelliJ plugins?

**Comment by JakeWharton** [comment]
No. It only uses the JDK 8 to compile but still targets Java 7 bytecode as
an output. Animal sniffer prevents us from using Java 8 APIs as well.

**Pull request merged**